neo/Consensus/ConsensusService.cs

```diff
@@ -177,30 +222,44 @@ private void OnPersistCompleted(Block block)
             InitializeConsensus(0);
         }
 
-        private void OnPrepareRequestReceived(ConsensusPayload payload, PrepareRequest message)
+        private void OnPrepareRequestReceived(ConsensusPayload payload, PrepareRequest message, bool regenerationCall = false)
         {
             if (context.State.HasFlag(ConsensusState.RequestReceived)) return;
             if (payload.ValidatorIndex != context.PrimaryIndex) return;
             Log($"{nameof(OnPrepareRequestReceived)}: height={payload.BlockIndex} view={message.ViewNumber} index={payload.ValidatorIndex} tx={message.TransactionHashes.Length}");
-            if (!context.State.HasFlag(ConsensusState.Backup)) return;
+            if (!context.State.HasFlag(ConsensusState.Backup) && !regenerationCall) return;
+
             if (payload.Timestamp <= context.Snapshot.GetHeader(context.PrevHash).Timestamp || payload.Timestamp > DateTime.UtcNow.AddMinutes(10).ToTimestamp())
             {
-                Log($"Timestamp incorrect: {payload.Timestamp}", LogLevel.Warning);
+                Log($"{nameof(OnPrepareRequestReceived)}: Timestamp incorrect: {payload.Timestamp}", LogLevel.Warning);
                 return;
             }
+
             context.State |= ConsensusState.RequestReceived;
             context.Timestamp = payload.Timestamp;
             context.Nonce = message.Nonce;
             context.NextConsensus = message.NextConsensus;
             context.TransactionHashes = message.TransactionHashes;
             context.Transactions = new Dictionary<UInt256, Transaction>();
-            byte[] hashData = context.MakeHeader().GetHashData();
-            if (!Crypto.Default.VerifySignature(hashData, message.Signature, context.Validators[payload.ValidatorIndex].EncodePoint(false))) return;
-            for (int i = 0; i < context.Signatures.Length; i++)
-                if (context.Signatures[i] != null)
-                    if (!Crypto.Default.VerifySignature(hashData, context.Signatures[i], context.Validators[i].EncodePoint(false)))
-                        context.Signatures[i] = null;
-            context.Signatures[payload.ValidatorIndex] = message.Signature;
+            context.PreparePayload = payload;
+            context.SignedPayloads[payload.ValidatorIndex] = message.PrepReqSignature;
+            context.FinalSignatures[payload.ValidatorIndex] = message.FinalSignature;
+
+            if (!CheckPrimaryPayloadSignature(payload))
+            {
+                context.SignedPayloads[payload.ValidatorIndex] = null;
+                return;
```

Comment:
`context.FinalSignatures[payload.ValidatorIndex] = null` ?

Reply:
@shargon, but, on the other hand, null is easy to check!! ehauehauea
Thus, if you want to cover this case I think it is good to add this check, @shargon.

In general, we tried to insert all types of verification and checks. In the future we can simplify some of them for speed ups.